massa-models/src/block_id.rs

```diff
@@ -86,74 +136,132 @@ impl FromStr for BlockId {
+                let data = chars.collect::<String>();
+                let decoded_bs58_check = bs58::decode(data)
+                    .with_check(None)
+                    .into_vec()
+                    .map_err(|_| ModelsError::BlockIdParseError)?;
+                let u64_deserializer = U64VarIntDeserializer::new(Included(0), Included(u64::MAX));
+                let (rest, version) = u64_deserializer
+                    .deserialize::<DeserializeError>(&decoded_bs58_check[..])
+                    .map_err(|_| ModelsError::BlockIdParseError)?;
+                match version {
+                    0 => Ok(Self(Hash::from_bytes(
+                        rest.try_into()
+                            .map_err(|_| ModelsError::BlockIdParseError)?,
+                    ))),
+                    _ => Err(ModelsError::BlockIdParseError),
+                }
+            }
+            _ => Err(ModelsError::BlockIdParseError),
+        }
     }
 }
 
 /// Serializer for `BlockId`
 #[derive(Default, Clone)]
-pub struct BlockIdSerializer;
+pub struct BlockIdSerializer {
+    version_serializer: U64VarIntSerializer,
+}
 
 impl BlockIdSerializer {
     /// Creates a new serializer for `BlockId`
     pub fn new() -> Self {
-        Self
+        Self {
+            version_serializer: U64VarIntSerializer::new(),
+        }
+    }
+}
+
+impl Serializer<BlockId> for BlockIdSerializer {
+    fn serialize(&self, value: &BlockId, buffer: &mut Vec<u8>) -> Result<(), SerializeError> {
+        self.version_serializer
+            .serialize(&value.get_version(), buffer)?;
+        match value {
+            BlockId::BlockIdV0(block_id) => self.serialize(block_id, buffer),
+        }
     }
 }
 
+#[transition::impl_version(versions("0"), structures("BlockId"))]
 impl Serializer<BlockId> for BlockIdSerializer {
     fn serialize(&self, value: &BlockId, buffer: &mut Vec<u8>) -> Result<(), SerializeError> {
-        buffer.extend(value.to_bytes());
+        buffer.extend(value.0.to_bytes());
         Ok(())
     }
 }
 
 /// Deserializer for `BlockId`
-#[derive(Default, Clone)]
+#[derive(Clone)]
 pub struct BlockIdDeserializer {
     hash_deserializer: HashDeserializer,
+    version_deserializer: U64VarIntDeserializer,
+}
+
+impl Default for BlockIdDeserializer {
+    fn default() -> Self {
+        Self::new()
+    }
 }
 
 impl BlockIdDeserializer {
     /// Creates a new deserializer for `BlockId`
     pub fn new() -> Self {
         Self {
             hash_deserializer: HashDeserializer::new(),
+            version_deserializer: U64VarIntDeserializer::new(Included(0), Included(u64::MAX)),
+        }
+    }
+}
+
+impl Deserializer<BlockId> for BlockIdDeserializer {
+    fn deserialize<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
+        &self,
+        buffer: &'a [u8],
+    ) -> IResult<&'a [u8], BlockId, E> {
+        if buffer.len() < 2 {
+            return Err(nom::Err::Error(E::from_error_kind(buffer, ErrorKind::Eof)));
+        }
```

Comment:
Not sure why length is checked at a raw 2 here, the component coming first is version serialized as a var int. Same goes for `OperationIdDeserializer`.

Reply:
We just check that we have at least a byte of version and something else

Reply:
Can we put a comment explaining that?

Reply:
Done.